tensorflow/contrib/eager/python/examples/generative_examples/dcgan.ipynb

```diff
@@ -1,324 +1,428 @@
         "colab_type": "code",
-        "id": "bkOfJxk5j5Hi"
-      },
-      "outputs": [],
-      "source": [
-        "class Discriminator(tf.keras.Model):\n",
-        "  def __init__(self):\n",
-        "    super(Discriminator, self).__init__()\n",
-        "    self.conv1 = tf.keras.layers.Conv2D(64, (5, 5), strides=(2, 2), padding='same')\n",
-        "    self.conv2 = tf.keras.layers.Conv2D(128, (5, 5), strides=(2, 2), padding='same')\n",
-        "    self.dropout = tf.keras.layers.Dropout(0.3)\n",
-        "    self.flatten = tf.keras.layers.Flatten()\n",
-        "    self.fc1 = tf.keras.layers.Dense(1)\n",
-        "\n",
-        "  def call(self, x, training=True):\n",
-        "    x = tf.nn.leaky_relu(self.conv1(x))\n",
-        "    x = self.dropout(x, training=training)\n",
-        "    x = tf.nn.leaky_relu(self.conv2(x))\n",
-        "    x = self.dropout(x, training=training)\n",
-        "    x = self.flatten(x)\n",
-        "    x = self.fc1(x)\n",
-        "    return x"
-      ]
+        "colab": {}
+      },
+      "cell_type": "code",
+      "source": [
+        "def make_discriminator_model():\n",
+        "   return tf.keras.Sequential([\n",
+        "      tf.keras.layers.Conv2D(64, (5, 5), strides=(2, 2), padding='same'),\n",
+        "      tf.keras.layers.LeakyReLU(),\n",
+        "      tf.keras.layers.Dropout(0.3),\n",
+        "      \n",
+        "      tf.keras.layers.Conv2D(128, (5, 5), strides=(2, 2), padding='same'),\n",
+        "      tf.keras.layers.LeakyReLU(),\n",
+        "      tf.keras.layers.Dropout(0.3),\n",
+        "       \n",
+        "      tf.keras.layers.Flatten(),\n",
+        "      tf.keras.layers.Dense(1)\n",
+        "     ])"
+      ],
+      "execution_count": 0,
+      "outputs": []
     },
     {
+      "metadata": {
+        "colab_type": "code",
+        "id": "gDkA05NE6QMs",
+        "colab": {}
+      },
       "cell_type": "code",
+      "source": [
+        "generator = make_generator_model()\n",
+        "discriminator = make_discriminator_model()"
+      ],
       "execution_count": 0,
+      "outputs": []
+    },
+    {
       "metadata": {
-        "colab": {},
-        "colab_type": "code",
-        "id": "gDkA05NE6QMs"
+        "colab_type": "text",
+        "id": "6TSZgwc2BUQ-"
       },
-      "outputs": [],
+      "cell_type": "markdown",
       "source": [
-        "generator = Generator()\n",
-        "discriminator = Discriminator()"
+        "\n",
+        "This model takes about ~30 seconds per epoch to train on a single Tesla K80 on Colab, as of July 2018. Eager execution can sometimes be slower than executing the equivalent graph due to overheads of interpreting Python code. By using [tf.contrib.eager.defun](https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun) to create graph functions, we get a ~10 secs/epoch performance boost. This way we get the best of both eager execution (easier for debugging) and graph mode (better performance)."
```

Comment:
Optional: Instead of `defun`ing the model calls, did you try `defun`-ing the whole training step? That might yield some good improvements too (and will be simpler)?

So instead of:

```python
def train(dataset, epochs, noise_dim):
  for epoch in range(epochs):
    for images in dataset:
      ...
```

You could have:

```python
def train_step(images):
   noise = tf.random_normal(...)
   with tf.GradientTape() as tape:
      ...
   

def train(dataset, epochs):
  for epoch in epochs:
    for images in dataset:
      train_step(images)
    # All that display stuff
```

Then for faster training you could do:
```python
train_step = tf.contrib.eager.defun(train_step)
```

Reply:
Updated to defun on train_steps instead of model calls. Performance boost is better at ~20 secs / epoch instead of ~10 secs / epoch (defun on model calls). Thanks!